src/transaction/duck_transaction_manager.cpp

```diff
@@ -422,6 +422,10 @@ idx_t DuckTransactionManager::GetCatalogVersion(Transaction &transaction_p) {
 void DuckTransactionManager::PushCatalogEntry(Transaction &transaction_p, duckdb::CatalogEntry &entry,
                                               duckdb::data_ptr_t extra_data, duckdb::idx_t extra_data_size) {
 	auto &transaction = transaction_p.Cast<DuckTransaction>();
+	if (!db.IsSystem() && !db.IsTemporary() && transaction.IsReadOnly()) {
+		throw InternalException("Attempting to do catalog changes on a transaction that is read-only - "
```

Comment:
This duplicates the check in `DuckTransactionManager::CommitTransaction`. The advantage of having the check here as well is that it allows detecting the problem at the point where things are actually going wrong (makes it more convenient when examining core dumps).